File: Tile.cpp
#include <algorithm> //std::reverse
#include <iostream>
#include <string>

#include "Debug.h"
#include "Tile.h"

Tile::Tile(){
	this->hasMonastery = false;
	this->roadsEnd = false;
	this->citiesAreIndependent = false;
	this->hasShield = false;
	this->edges = std::vector<int>(NUM_TILE_EDGES);
	this->tileNodes = std::vector<GraphNode*>(NUM_TILE_NODES);
	this->tileType = UNKNOWN_TILE_TYPE;
	this->rotations = 0;
	return;
}

Tile::Tile(bool bHasMonastery, bool bRoadsEnd, bool bCitiesAreIndependent, bool bHasShield, std::vector<int> Edges)
{
	this->hasMonastery = bHasMonastery;
	this->roadsEnd = bRoadsEnd;
	this->citiesAreIndependent = bCitiesAreIndependent;
	this->hasShield = bHasShield;
	this->edges = Edges;
	this->tileType = UNKNOWN_TILE_TYPE;
	this->rotations = 0;
	return;
}

Tile::Tile(bool bHasMonastery, bool bRoadsEnd, bool bCitiesAreIndependent, bool bHasShield, std::vector<int> Edges, char cTileType)
{
	this->hasMonastery = bHasMonastery;
	this->roadsEnd = bRoadsEnd;
	this->citiesAreIndependent = bCitiesAreIndependent;
	this->hasShield = bHasShield;
	this->edges = Edges;
	this->tileNodes.resize(NUM_TILE_NODES);
	this->tileType = cTileType;
	this->rotations = 0;

	//Count up number of road edges and city edges, for use later
	int numRoads = 0, numCities = 0;
	for (unsigned int edgeIndex = 0; edgeIndex < this->edges.size(); edgeIndex++) {
		if (edges[edgeIndex] == TerrainType::City) numCities++;
		else if (edges[edgeIndex] == TerrainType::Road) numRoads++;
	}

	//Handles the creation of each node, one edge at a time, as well as identifying features ending
	for (unsigned int edgeIndex = 0; edgeIndex < this->edges.size(); edgeIndex++) {
		//First and third nodes are always plains if the edge is not a city, and a city if the edge is a city. Second node is always equal to edge type.
		tileNodes[edgeIndex * (NODES_PER_EDGE)] = (edges[edgeIndex] == TerrainType::City) ? new GraphNode(TerrainType::City) : new GraphNode(TerrainType::Plains);
		tileNodes[edgeIndex * (NODES_PER_EDGE) + 1] = new GraphNode(edges[edgeIndex]);
		tileNodes[edgeIndex * (NODES_PER_EDGE) + 2] = (edges[edgeIndex] == TerrainType::City) ? new GraphNode(TerrainType::City) : new GraphNode(TerrainType::Plains);

		if (this->roadsEnd) tileNodes[edgeIndex * (NODES_PER_EDGE)+1]->featureEnd = true;

		if (edges[edgeIndex] == TerrainType::City && (this->citiesAreIndependent || numCities == 1)) {
			for (unsigned int subIndex = 0; subIndex < NODES_PER_EDGE; subIndex++) {
				tileNodes[edgeIndex * NODES_PER_EDGE + subIndex]->featureEnd = true;
			}
		}
		else if (edges[edgeIndex] == TerrainType::Road && (this->roadsEnd || numRoads == 1)) {
			tileNodes[edgeIndex * NODES_PER_EDGE + 1]->featureEnd = true;
		}
	}
	tileNodes[NUM_TILE_NODES - 1] = this->hasMonastery ? new GraphNode(TerrainType::Monastery) : new GraphNode(TerrainType::None);

	//Handles linking all adjacent nodes within a tile that need to be linked
	for (unsigned int nodeIndex = 0; nodeIndex < this->tileNodes.size() - 1; nodeIndex++) { //-1 because this part should handle only outer nodes
		if (this->citiesAreIndependent && this->tileNodes[nodeIndex]->nodeType == TerrainType::City && (nodeIndex + 1) % NODES_PER_EDGE == 0) { // Identifies corners of tiles
			continue;
		}

		GraphNode* currNode = this->tileNodes[nodeIndex];
		GraphNode* nextNode = this->tileNodes[(nodeIndex + 1) % (NUM_TILE_NODES - 1)];

		if (currNode->nodeType == nextNode->nodeType) {
			currNode->connectedNodes.push_back(nextNode);
			nextNode->connectedNodes.push_back(currNode);
		}
	}
	
	if (numRoads == 2) {
		for (unsigned int nodeIndex = 1; nodeIndex < this->tileNodes.size() - 1; nodeIndex += NODES_PER_EDGE) {
			if (tileNodes[nodeIndex]->nodeType == TerrainType::Road) {
				for (unsigned int targetNodeIndex = nodeIndex + NODES_PER_EDGE; targetNodeIndex < this->tileNodes.size(); targetNodeIndex += NODES_PER_EDGE) {
					if (tileNodes[targetNodeIndex]->nodeType == TerrainType::Road) {
						tileNodes[nodeIndex]->connectedNodes.push_back(tileNodes[targetNodeIndex]);
						tileNodes[targetNodeIndex]->connectedNodes.push_back(tileNodes[nodeIndex]);
						goto roadStop; //Break all the way out of nested loop
					}
				}
			}
		}
	}

roadStop:
	if (numCities == 2 && !this->citiesAreIndependent) {
		for (unsigned int nodeIndex = 1; nodeIndex < (this->tileNodes.size() - 1)/2; nodeIndex += NODES_PER_EDGE) {
			if (tileNodes[nodeIndex]->nodeType == TerrainType::City && tileNodes[nodeIndex + NODES_PER_EDGE * 2]->nodeType == TerrainType::City) {
				tileNodes[nodeIndex]->connectedNodes.push_back(tileNodes[nodeIndex + NODES_PER_EDGE * 2]);
				tileNodes[nodeIndex + NODES_PER_EDGE * 2]->connectedNodes.push_back(tileNodes[nodeIndex]);
				break;
			}
		}
	}
	
	return;
}

Tile::~Tile() {

	return;
}

void Tile::RotateClockwise(int rotations) {
	rotations %= NUM_TILE_EDGES; //4 rotations is equivalent to zero, 5 to 1, etc.

	if (rotations == 0) return;
	else if (rotations < 0) rotations += NUM_TILE_EDGES; //modulo operator can give negative numbers, add to num_tile_edges to offset this behavior

	this->rotations = (this->rotations + rotations) % NUM_TILE_EDGES; //no need to check for negative, since this->rot always >= 0 and rotations is always >= 0 by now.

	std::rotate(edges.begin(), edges.end() - rotations, edges.end());

	std::rotate(tileNodes.begin(), tileNodes.end() - 1 - (rotations*NODES_PER_EDGE), tileNodes.end() - 1); //Don't include the last vector item (tile-center), it should never move!

	return;
}


std::vector<int> Tile::GetEdges()
{
	return this->edges;
}

bool Tile::GetHasMonastery()
{
	return this->hasMonastery;
}

bool Tile::GetRoadsEnd()
{
	return this->roadsEnd;
}

bool Tile::GetCitiesAreIndependent()
{
	return this->citiesAreIndependent;
}

bool Tile::GetHasShield()
{
	return this->hasShield;
}

std::string ConvertEdgeEnumToString(int edgeValue) {
	switch (edgeValue) {
	case TerrainType::Plains: return "Plains";
	case TerrainType::Road: return "Road";
	case TerrainType::City: return "City";
	default: return "";
	}
}

void Tile::PrintTileInformation(bool printVerbose) {
	if (printVerbose) {
		std::cout << "Tile Type: " << tileType << std::endl
			<< "Has Monastery: " << ((hasMonastery) ? "Yes" : "No") << std::endl
			<< "Roads End Here: " << ((roadsEnd) ? "Yes" : "No") << std::endl
			<< "Independent Cities: " << ((citiesAreIndependent) ? "Yes" : "No") << std::endl
			<< "Has Shield: " << ((hasShield) ? "Yes" : "No") << std::endl
			<< "North: " << ConvertEdgeEnumToString(edges[0]) << std::endl
			<< "East: " << ConvertEdgeEnumToString(edges[1]) << std::endl
			<< "South: " << ConvertEdgeEnumToString(edges[2]) << std::endl
			<< "West: " << ConvertEdgeEnumToString(edges[3]) << std::endl;
	}
	else {
		std::cout << tileType << " " << hasMonastery << " " << roadsEnd << " " << citiesAreIndependent << " " << hasShield
			<< " " << edges[0] << " " << edges[1] << " " << edges[2] << " " << edges[3] << std::endl;
	}

	return;
}

void Tile::PrintTileNodeInformation(bool printVerbose) {
	if (printVerbose) {
		std::cout << "NW " << std::endl;
		for (unsigned int Index = 0; Index < tileNodes.size(); Index++) {
			std::cout << tileNodes[Index]->nodeType << " -> ";
			
			for (GraphNode* node : tileNodes[Index]->connectedNodes) {
				std::cout << node->nodeType << ", ";
			}

			std::cout << std::endl;
		}
	}
	else {
		std::cout << "NW ";
		for (unsigned int Index = 0; Index < tileNodes.size(); Index++) {
			if (Index % NODES_PER_EDGE == 0) std::cout << " ";
			std::cout << tileNodes[Index]->nodeType;
		}
		std::cout << std::endl;
	}
}

bool Tile::operator==(const Tile& tile) {
	if (this->tileType != UNKNOWN_TILE_TYPE) {
		return (this->tileType == tile.tileType);
	}
	
	//If we do not control the input tiles, then we will need to /always/ use the following code
	//which more thoroughly checks if two tiles are equal. As long as we know the tile types are
	//valid then we can simply compare those, as above.

	if (this->hasMonastery == tile.hasMonastery &&
		this->roadsEnd == tile.roadsEnd &&
		this->citiesAreIndependent == tile.citiesAreIndependent &&
		this->hasShield == tile.hasShield) {

		std::vector<int> storedEdges = this->edges;

		for (int RotationAttempt = 0; RotationAttempt < NUM_TILE_EDGES; RotationAttempt++) {
			if (this->edges == tile.edges) {
				this->edges = storedEdges; //Restore edges to their original
				return true;
			}
			else this->RotateClockwise(1);
		}

		this->edges = storedEdges; //In case we fall through
	}

	return false;
}

File: Tile.java
import javafx.scene.transform.Rotate;

import javax.smartcardio.TerminalFactory;
import java.util.*;
import java.util.ArrayList;

public class Tile {
	private static final int EDGES_PER_TILE = 4;
	private static final int NODES_PER_EDGE = 3;
	private static int identify = 0;
	
	int ID,
		rotations = 0;
	Integer[] edgeValues;
	Edge[] edges = new Edge[
			EDGES_PER_TILE];
	Node middle;
	char tileType;
	boolean hasDen,
		 	trailsEnd,
			lakesAreIndependent;
	int animalType;

	static void resetTileIdentify(){
		identify = 0;
	}

	public Tile(Tile tileToCopy){
		Tile newTile = new Tile(tileToCopy.hasDen, tileToCopy.trailsEnd, tileToCopy.lakesAreIndependent, tileToCopy.animalType, tileToCopy.edgeValues, tileToCopy.tileType);
		this.hasDen = newTile.hasDen;
		this.trailsEnd = newTile.trailsEnd;
		this.lakesAreIndependent = newTile.lakesAreIndependent;
		this.tileType = newTile.tileType;
		this.rotations = newTile.rotations;
		this.edgeValues = newTile.edgeValues;
		this.animalType = newTile.animalType;
		this.edges = newTile.edges;
		this.middle = newTile.middle;
		this.ID = tileToCopy.ID % 77;
	}
	
	public Tile(Tile tileToCopy, boolean unrotate){
		Tile newTile = new Tile(tileToCopy.hasDen, tileToCopy.trailsEnd, tileToCopy.lakesAreIndependent, tileToCopy.animalType, tileToCopy.edgeValues, tileToCopy.tileType);
		
		Tile.printTile(newTile);
		if (unrotate) newTile.rotateClockwise(-tileToCopy.rotations);
		Tile.printTile(newTile);
		this.hasDen = newTile.hasDen;
		this.trailsEnd = newTile.trailsEnd;
		this.lakesAreIndependent = newTile.lakesAreIndependent;
		this.tileType = newTile.tileType;
		this.rotations = 0;
		this.edgeValues = newTile.edgeValues;
		this.animalType = newTile.animalType;
		this.edges = newTile.edges;
		this.middle = newTile.middle;
		this.ID = tileToCopy.ID % 77;
		Tile.printTile(this);
	}
	
	

	public Tile(boolean hasDen, boolean trailsEnd, boolean lakesAreIndependent, int animalType, Integer[] edgeValues, char tileType){
		this.hasDen = hasDen;
		this.trailsEnd = trailsEnd;
		this.lakesAreIndependent = lakesAreIndependent;
		this.tileType = tileType;
		this.rotations = 0;
		this.edgeValues = edgeValues;
		this.animalType = animalType;
		this.ID = identify++ % 77;

		//Count number of trail + city edges
		int numTrails = 0, numCities = 0;
		ArrayList<Integer> cityLocations = new ArrayList<Integer>();
		for (int edgeIndex = 0; edgeIndex < edgeValues.length; edgeIndex++){
			if (edgeValues[edgeIndex] == FeatureTypeEnum.Lake.toInt()){
				numCities++;
				cityLocations.add(edgeIndex);
			}
			else if (edgeValues[edgeIndex] == FeatureTypeEnum.Trail.toInt()){
				numTrails++;
			}
		}

		//Can't combine these two loops, this depends on numCities and numTrails.
		//This loop handles the creation of all the edges (and their composing nodes) for the tile.
		for (int edgeIndex = 0; edgeIndex < edgeValues.length; edgeIndex++){
			Node[] nodeBuffer = new Node[NODES_PER_EDGE];

			for (int nodeIndex = 0; nodeIndex < nodeBuffer.length; nodeIndex++) {
				switch (nodeIndex) {
					case 0:	// For first and third nodes, always field if not a city, otherwise always a city
					case 2:
						nodeBuffer[nodeIndex] = (edgeValues[edgeIndex] == FeatureTypeEnum.Lake.toInt()) ? new Node(FeatureTypeEnum.Lake.toChar(), this.ID) : new Node(FeatureTypeEnum.Jungle.toChar(), this.ID);
						break;
					case 1: // For second node, always equal to the type of edge it is on.
						if (edgeValues[edgeIndex] == FeatureTypeEnum.Lake.toInt())
							nodeBuffer[nodeIndex] = new Node(FeatureTypeEnum.Lake.toChar(), this.ID);
						else if (edgeValues[edgeIndex] == FeatureTypeEnum.Trail.toInt())
							nodeBuffer[nodeIndex] = new Node(FeatureTypeEnum.Trail.toChar(), this.ID);
						else if (edgeValues[edgeIndex] == FeatureTypeEnum.Jungle.toInt())
							nodeBuffer[nodeIndex] = new Node(FeatureTypeEnum.Jungle.toChar(), this.ID);
						else
							System.out.println("Error with tile import: invalid edge value");
						break;
					default:
						System.out.println("Node buffer should be out of bounds");
						break;
				}
			}

			//Handle some of the edge cases down here. Certainly can be incorporated above, but makes it much harder to read
			if (edgeValues[edgeIndex] == FeatureTypeEnum.Lake.toInt() && (this.lakesAreIndependent || numCities == 1)){
				for (int nodeIndex = 0; nodeIndex < nodeBuffer.length; nodeIndex++){
					nodeBuffer[nodeIndex].featureType = FeatureTypeEnum.Shore;
				}
			}
			else if (edgeValues[edgeIndex] == FeatureTypeEnum.Trail.toInt() && (this.trailsEnd || numTrails == 1)){
				nodeBuffer[1].featureType = FeatureTypeEnum.TrailEnd;
			}

			this.edges[edgeIndex] = new Edge(nodeBuffer);
		}

		if (this.hasDen) {
			this.middle = new Node(FeatureTypeEnum.Den.toChar(), this.ID);

			for (Edge edge : this.edges){
				if (edge.nodes[1].featureType == FeatureTypeEnum.Jungle){
					addEachNodeAsNeighbor(middle, edge.nodes[1]);
					break;
				}
			}
		}
		else this.middle = new Node(FeatureTypeEnum.None.toChar(), this.ID); //none

		//Now we need to handle city edge cases
		//numCities == 1 case is already handled above
		if (numCities == 2 && !this.lakesAreIndependent){ //Independent lakes are already handled above
			int firstLakeLoc = cityLocations.get(0), secondLakeLoc = cityLocations.get(1);
			if (secondLakeLoc - firstLakeLoc == 2){ //In this case, the lakes oppose each other on the tile
				for (int nodeIndex = 0; nodeIndex < NODES_PER_EDGE; nodeIndex++){
					int antiNodeIndex = NODES_PER_EDGE - 1 - nodeIndex; //Used to mirror the nodeIndex on edges opposite each other

					if (nodeIndex != 1) { //Prevents the type of the middle node from being changed
						edges[firstLakeLoc].nodes[nodeIndex].featureType = FeatureTypeEnum.Shore;
						edges[secondLakeLoc].nodes[antiNodeIndex].featureType = FeatureTypeEnum.Shore;
					}
					//Then connect opposing nodes
					addEachNodeAsNeighbor(edges[firstLakeLoc].nodes[nodeIndex], edges[secondLakeLoc].nodes[antiNodeIndex]);
				}
			}
			else{ //Otherwise, the two city edges are adjacent on the tile
				if (secondLakeLoc - firstLakeLoc == 3) { int swap = firstLakeLoc; firstLakeLoc = secondLakeLoc; secondLakeLoc = swap; }
				//These [inner nodes] will be automatically linked later by the adjacency linker
				edges[firstLakeLoc].nodes[2].featureType = FeatureTypeEnum.InnerShore;
				edges[secondLakeLoc].nodes[0].featureType = FeatureTypeEnum.InnerShore;

				//Outer node set up. Need to be linked here.
				edges[firstLakeLoc].nodes[0].featureType = FeatureTypeEnum.Shore;
				edges[secondLakeLoc].nodes[2].featureType = FeatureTypeEnum.Shore;

				addEachNodeAsNeighbor(edges[firstLakeLoc].nodes[0], edges[secondLakeLoc].nodes[2]);
				addEachNodeAsNeighbor(edges[firstLakeLoc].nodes[1], edges[secondLakeLoc].nodes[1]);
			}
		}
		else if (numCities == 3){
			for (int edgeIndex = 0; edgeIndex < edges.length; edgeIndex++){
				//This sets up the true wall neighboring the field
				if(edges[edgeIndex].nodes[2].featureType == FeatureTypeEnum.Lake && edges[(edgeIndex + 1) % edges.length].nodes[2].featureType == FeatureTypeEnum.Jungle){
					edges[edgeIndex].nodes[2].featureType = FeatureTypeEnum.Shore;
					edges[(edgeIndex + 2) % edges.length].nodes[0].featureType = FeatureTypeEnum.Shore;

					addEachNodeAsNeighbor(edges[edgeIndex].nodes[2], edges[(edgeIndex + 2) % edges.length].nodes[0]);

					//Connects inner city nodes
					addEachNodeAsNeighbor(edges[edgeIndex].nodes[1], edges[(edgeIndex + 2) % edges.length].nodes[1]);
					addEachNodeAsNeighbor(edges[edgeIndex].nodes[1], edges[(edgeIndex + 3) % edges.length].nodes[1]);
					addEachNodeAsNeighbor(edges[(edgeIndex + 3) % edges.length].nodes[1], edges[(edgeIndex + 2) % edges.length].nodes[1]);
				}

				//This sets up the inner connections to be linked later
				if((edgeValues[edgeIndex] == FeatureTypeEnum.Lake.toInt()) && edgeValues[(edgeIndex + 1) % edges.length] == FeatureTypeEnum.Lake.toInt()){
					edges[edgeIndex].nodes[2].featureType = FeatureTypeEnum.InnerShore;
					edges[(edgeIndex + 1) % edges.length].nodes[0].featureType = FeatureTypeEnum.InnerShore;
				}
			}
		}
		else if (numCities == 4){
			//Will be linked later by adjacency linker
			for (int edgeIndex = 0; edgeIndex < edges.length; edgeIndex++) {
				edges[edgeIndex].nodes[0].featureType = FeatureTypeEnum.InnerShore;
				edges[edgeIndex].nodes[2].featureType = FeatureTypeEnum.InnerShore;

				addEachNodeAsNeighbor(edges[edgeIndex].nodes[1], (edges[(edgeIndex + 1) % EDGES_PER_TILE]).nodes[1]);
			}
		}

		/*Adjacency linker*/
		for (int edgeIndex = 0; edgeIndex < edges.length; edgeIndex++){
			for (int nodeIndex = 0; nodeIndex <
					NODES_PER_EDGE - 1; nodeIndex++) { //We can only loop for 1st two nodes on an edge. Third node needs to be handled specially
				//if (edges[edgeIndex].nodes[nodeIndex].featureType == edges[edgeIndex].nodes[nodeIndex + 1].featureType || 					//If same feature type
				//		edges[edgeIndex].nodes[nodeIndex].featureType.isShoreToLake(edges[edgeIndex].nodes[nodeIndex + 1].featureType)) {	//Or a wall to city connection
				//	addEachNodeAsNeighbor(edges[edgeIndex].nodes[nodeIndex], edges[edgeIndex].nodes[nodeIndex + 1]);
				//}

				//Just link all adjacent nodes if they are the same feature.
				if (edges[edgeIndex].nodes[nodeIndex].featureType.isSameFeature(edges[edgeIndex].nodes[nodeIndex+1].featureType)){
					addEachNodeAsNeighbor(edges[edgeIndex].nodes[nodeIndex], edges[edgeIndex].nodes[nodeIndex + 1]);
				}
			}

			//3rd node handler
			if (this.lakesAreIndependent &&
					edges[edgeIndex].nodes[2].featureType == FeatureTypeEnum.Shore &&
					edges[(edgeIndex + 1) % edges.length].nodes[0].featureType == FeatureTypeEnum.Shore){
				continue;
			}

			addEachNodeAsNeighbor(edges[edgeIndex].nodes[2], edges[(edgeIndex + 1) % edges.length].nodes[0]);
		}

		//Trail handler
		trailLoop:
		if (numTrails == 2){
			for (int edgeIndex = 0; edgeIndex < edges.length; edgeIndex++){
				if (edges[edgeIndex].nodes[1].featureType == FeatureTypeEnum.Trail){
					for (int targetEdgeIndex = edgeIndex + 1; targetEdgeIndex < edges.length; targetEdgeIndex++){
						if (edges[targetEdgeIndex].nodes[1].featureType == FeatureTypeEnum.Trail){
							//Connect the trails
							addEachNodeAsNeighbor(edges[edgeIndex].nodes[1], edges[targetEdgeIndex].nodes[1]);
							//Consider fields next to trail for connections
							if (numCities == 1){
								int lowerIndex = cityLocations.get(0) - 1;
								if (lowerIndex < 0) lowerIndex += EDGES_PER_TILE;
								addEachNodeAsNeighbor(edges[lowerIndex].nodes[2], edges[(cityLocations.get(0) + 1) % EDGES_PER_TILE].nodes[0]);
							}
							else if (numCities == 2){
								int lowerIndex = cityLocations.get(0) - 1;
								if (lowerIndex < 0) lowerIndex += EDGES_PER_TILE;
								addEachNodeAsNeighbor(edges[lowerIndex].nodes[2], edges[(cityLocations.get(1) + 1) % EDGES_PER_TILE].nodes[0]);
							}
							break trailLoop;
						}
					}
				}
			}
		}
		else if (numTrails == 3 && numCities > 0){
			int lowerIndex = cityLocations.get(0) - 1;
			if (lowerIndex < 0) lowerIndex += EDGES_PER_TILE;
			addEachNodeAsNeighbor(edges[lowerIndex].nodes[2], edges[(cityLocations.get(0) + 1) % EDGES_PER_TILE].nodes[0]);
		}

		if(this.tileType == 'H'){
			this.printNodeHashes();
		}
	}

	private void addEachNodeAsNeighbor(Node nodeA, Node nodeB){
		nodeA.neighbors.add(nodeB);
		nodeB.neighbors.add(nodeA);
	}

	boolean isEqual(Tile other){
		//More robust checks do not need to be implemented as long as we control tileset input.
		return this.tileType == other.tileType;
	}


	void rotateClockwise(int rotations){
		rotations %= EDGES_PER_TILE;
		if (rotations == 0) return;
		else if (rotations < 0) rotations += EDGES_PER_TILE;

		this.rotations = (this.rotations + rotations) % EDGES_PER_TILE;
		
		Collections.rotate(Arrays.asList(edges), rotations);
		Collections.rotate(Arrays.asList(edgeValues), rotations);
		return;
	}

	void rotateAntiClockwise(int rotations){
		rotations %= EDGES_PER_TILE;
		if (rotations == 0) return;
		else if (rotations < 0) rotations += EDGES_PER_TILE;

		rotateClockwise(Math.abs(rotations - EDGES_PER_TILE));
	}

	public void printNodeHashes(){
		for (Edge edge : this.edges){
			for (Node node : edge.nodes){
				System.out.print("\n" + node.hashCode() + "(" + node.featureType + "): ");
				for (Node neighbor : node.neighbors){
					System.out.print(neighbor.hashCode() + "(" + neighbor.featureType + "), ");
				}
			}
		}
	}
    
    static void printTile(Tile toPrint){
        System.out.print(" ");
        for(int i = 0; i < 3; i++){
            System.out.print(toPrint.edges[1].nodes[i].featureType.toChar());
        }
        System.out.println("");
        System.out.println(toPrint.edges[0].nodes[2].featureType.toChar() + "   " + toPrint.edges[2].nodes[0].featureType.toChar());
        System.out.println(toPrint.edges[0].nodes[1].featureType.toChar() + "   " + toPrint.edges[2].nodes[1].featureType.toChar());
        System.out.println(toPrint.edges[0].nodes[0].featureType.toChar() + "   " + toPrint.edges[2].nodes[2].featureType.toChar());
        System.out.print(" ");
        for(int i = 2; i >= 0; i--){
            System.out.print(toPrint.edges[3].nodes[i].featureType.toChar());
        }
        System.out.println("");
    }
    
}